File: lib/results_page.dart
import 'package:flutter/material.dart';
import 'package:whatsinmyfood/components/scan_barcode.dart';
import 'package:whatsinmyfood/main.dart';
import 'food_lists.dart';
import 'components/global_variables.dart';

class ResultsPage extends StatelessWidget {
  final List<String> passedResults;
  final Function(String) updateScanResult;

  const ResultsPage({
    Key? key,
    required this.passedResults,
    required BuildContext context,
    required this.updateScanResult,
  }) : super(key: key);

  Map<String, List<String>> categorizeResults(
      List<String> passedResults,
      List<Map<String, List<String>>> keywordLists,
      List<String> lookingForThings) {
    // Convert lookingForThings to lowercase and replace spaces with hyphens
    List<String> lowercaseLookingForThings = lookingForThings
        .map((item) => item.toLowerCase().replaceAll(' ', '-'))
        .toList();

    // Convert passedResults to lowercase and replace spaces with hyphens
    List<String> lowercasePassedResults = passedResults
        .map((item) => item.toLowerCase().replaceAll(' ', '-'))
        .toList();

    Map<String, List<String>> categorizedResults = {};

    // Convert all keywords and elements in keyword lists to lowercase and replace spaces with hyphens
    List<Map<String, List<String>>> lowercaseKeywordLists =
        keywordLists.map((map) {
      Map<String, List<String>> lowercaseMap = {};
      map.forEach((key, value) {
        List<String> lowercaseList = value
            .map((element) => element.toLowerCase().replaceAll(' ', '-'))
            .toList();
        lowercaseMap[key.toLowerCase().replaceAll(' ', '-')] = lowercaseList;
      });
      return lowercaseMap;
    }).toList();

    // Add all keyword maps to categorizedResults
    for (var map in lowercaseKeywordLists) {
      categorizedResults.addAll(map);
    }

    // Debug print to check categorizedResults before filtering
    //print('Categorized results before filtering: $categorizedResults');

    // Filter the categorizedResults to include only the keys from lowercaseLookingForThings
    categorizedResults = Map.fromEntries(categorizedResults.entries
        .where((entry) => lowercaseLookingForThings.contains(entry.key)));

    // Debug print to check categorizedResults after filtering
    //print('Categorized results after filtering: $categorizedResults');

    print('Looking for things: $lowercaseLookingForThings');
    print('Passed results: $lowercasePassedResults');
    print('Keyword Lists: $lowercaseKeywordLists');

    for (String result in lowercasePassedResults) {
      for (var keywordMap in lowercaseKeywordLists) {
        keywordMap.forEach((keyword, list) {
          String lowercaseKeyword = keyword.toLowerCase().replaceAll(' ', '-');
          if (list.any((element) =>
              result.contains(element.toLowerCase().replaceAll(' ', '-')))) {
            if (lowercasePassedResults.contains(result)) {
              categorizedResults.putIfAbsent(lowercaseKeyword, () => []);
              categorizedResults[lowercaseKeyword]?.add(result);
            }
          }
        });
      }
    }
    passedResults.clear();
    print('Final categorizedResults: $categorizedResults');
    return categorizedResults;
  }

  @override
  Widget build(BuildContext context) {
    Map<String, List<String>> categorizedResults =
        categorizeResults(passedResults, keywordLists, lookingForThings);

    return Scaffold(
      appBar: AppBar(
        title: const Text('Results Page'),
      ),
      body: Column(
        children: [
          ElevatedButton(
            onPressed: () async {
              await Navigator.push(
                context,
                MaterialPageRoute(
                  builder: (context) => const HomePage(),
                ),
              );
            },
            style: ElevatedButton.styleFrom(
              backgroundColor: Colors.blue,
              foregroundColor: Colors.white,
            ),
            child: const Text(
              'Adjust filters',
              style: TextStyle(color: Colors.white),
            ),
          ),
          Expanded(
            child: ListView(
              children: categorizedResults.entries
                  .where((entry) =>
                      entry.value.isNotEmpty) // Filter out empty lists
                  .map((entry) {
                var keyword = entry.key;
                var matchingElements = entry.value;

                return ExpansionTile(
                  title: Text(keyword
                      .toUpperCase()
                      .replaceAll('-', ' ')), // title is here
                  initiallyExpanded: true,
                  children: [
                    ListView.builder(
                      shrinkWrap: true,
                      itemCount: matchingElements.length,
                      itemBuilder: (context, index) {
                        return ListTile(
                          title:
                              Text(matchingElements[index]), // sublist is here
                        );
                      },
                    ),
                  ],
                );
              }).toList(),
            ),
          ),
          Center(
            child: Column(
              mainAxisAlignment: MainAxisAlignment.center,
              children: [
                //Text("Looking for: $lookingForThings"),
                Text('YOUR RESULTS: $passedResults'),
                const SizedBox(
                    height: 16), // Add some spacing between the buttons
                ElevatedButton(
                  onPressed: () {
                    handleBarcodeScan(
                      context,
                      lookingForThings,
                      updateScanResult,
                    );
                  },
                  style: ElevatedButton.styleFrom(
                    backgroundColor: Colors.red,
                    foregroundColor: Colors.white,
                  ),
                  child: const Text(
                    'Scan again',
                    style: TextStyle(color: Colors.white),
                  ),
                ),
              ],
            ),
          ),
        ],
      ),
    );
  }
}

File: composeApp/src/commonMain/kotlin/org/whatsinmyfood/project/pages/results_page.kt
/*
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.widget.Button
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.button.MaterialButton
import com.google.android.material.card.MaterialCardView

class ResultsPageActivity : AppCompatActivity() {

    private lateinit var passedResults: List<String>
    private val keywordLists: List<Map<String, List<String>>> = listOf() // Initialize with actual data
    private val lookingForThings: List<String> = listOf() // Initialize with actual data

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_results_page)

        passedResults = intent.getStringArrayListExtra("passedResults") ?: listOf()

        val categorizedResults = categorizeResults(passedResults, keywordLists, lookingForThings)

        // Set up RecyclerView with categorized results
        val recyclerView = findViewById<RecyclerView>(R.id.recyclerViewResults)
        recyclerView.layoutManager = LinearLayoutManager(this)
        recyclerView.adapter = ResultsAdapter(categorizedResults)

        // Handle adjust filters button
        val adjustFiltersButton = findViewById<Button>(R.id.adjustFiltersButton)
        adjustFiltersButton.setOnClickListener {
            val intent = Intent(this, HomePageActivity::class.java)
            startActivity(intent)
        }

        // Handle scan again button
        val scanAgainButton = findViewById<Button>(R.id.scanAgainButton)
        scanAgainButton.setOnClickListener {
            // Implement scanning logic here (e.g., launching a barcode scanner activity)
            handleBarcodeScan(this, {}, passedResults)
        }
    }

    private fun categorizeResults(
        passedResults: List<String>,
        keywordLists: List<Map<String, List<String>>>,
        lookingForThings: List<String>
    ): Map<String, List<String>> {
        // Convert lookingForThings and passedResults to lowercase and replace spaces with hyphens
        val lowercaseLookingForThings = lookingForThings.map { it.lowercase().replace(" ", "-") }
        val lowercasePassedResults = passedResults.map { it.lowercase().replace(" ", "-") }

        val categorizedResults = mutableMapOf<String, MutableList<String>>()

        // Convert all keywords and elements in keyword lists to lowercase and replace spaces with hyphens
        val lowercaseKeywordLists = keywordLists.map { map ->
            map.mapKeys { it.key.lowercase().replace(" ", "-") }
                .mapValues { it.value.map { elem -> elem.lowercase().replace(" ", "-") } }
        }

        // Add all keyword maps to categorizedResults
        lowercaseKeywordLists.forEach { map ->
            map.forEach { (key, value) ->
                categorizedResults[key] = mutableListOf()
            }
        }

        // Filter the categorizedResults to include only the keys from lowercaseLookingForThings
        val filteredResults = categorizedResults.filterKeys { lowercaseLookingForThings.contains(it) }

        // Iterate through lowercasePassedResults and keyword lists to categorize results
        lowercasePassedResults.forEach { result ->
            lowercaseKeywordLists.forEach { keywordMap ->
                keywordMap.forEach { (keyword, list) ->
                    if (list.any { result.contains(it) }) {
                        if (lowercasePassedResults.contains(result)) {
                            filteredResults[keyword]?.add(result)
                        }
                    }
                }
            }
        }

        return filteredResults.filter { it.value.isNotEmpty() } // Return only non-empty results
    }
}
 */